File: linked-list/merge-two-sorted-lists/enrique/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        result_iter = result
        while l1 and l2:
            a = l1.val
            b = l2.val
            next_node = None
            if a <= b:
                next_node = ListNode(a)
                l1 = l1.next
            else:
                next_node = ListNode(b)
                l2 = l2.next

            result_iter.next = next_node
            result_iter = result_iter.next

        result_iter.next = l1 if l1 is not None else l2
        return result.next

File: linked-list/merge-two-sorted-lists/enrique/merge-two-sorted-lists.go
package enrique

// ListNode Definition for singly-linked list.
type ListNode struct {
    Val int
    Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	if  l1 == nil && l2 == nil {
		return nil
	}

	solutionRoot := &ListNode{}

	iterL1 := l1
	iterL2 := l2
	mergedSolutionIter := solutionRoot
	for iterL1 != nil && iterL2 != nil {
		if iterL1.Val < iterL2.Val {
			mergedSolutionIter.Val = iterL1.Val
			iterL1 = iterL1.Next
		} else {
			mergedSolutionIter.Val = iterL2.Val
			iterL2 = iterL2.Next
		}

		mergedSolutionIter.Next = &ListNode{}
		mergedSolutionIter = mergedSolutionIter.Next
	}

	if iterL1 != nil {
		mergedSolutionIter.Val = iterL1.Val
		mergedSolutionIter.Next = iterL1.Next
	}

	if iterL2 != nil {
		mergedSolutionIter.Val = iterL2.Val
		mergedSolutionIter.Next = iterL2.Next
	}

	return solutionRoot
}
